### PR TITLE
WIP: Experimental method to aggregate state effects

### PR DESCRIPTION
**Pull request by shigoel**
### Description:

Experimental method, inspired by `simp_arith`, to aggregate state effects.

### Testing:

What tests have been run? Did `make all` succeed for your changes? Was
conformance testing successful on an Aarch64 machine?

Yes

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Tactics/ArmConstr.lean [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+Author(s): Shilpi Goel
+-/
+
+/-
+Experimental method to aggregate state effects using reflection. This is
+inspired by `simp_arith`, especially the following files:
+
+`[lean] Init/Data/Nat/Linear.lean`
+`[lean] Meta/Tactic/LinearAith/Nat/Basic.lean`
+`[lean] Meta/Tactic/LinearAith/Nat/Simp.lean`
```
Comment (bollu) [inline Tactics/ArmConstr.lean:13]:
```suggestion
`[lean] Meta/Tactic/LinearAith/Nat/Simp.lean`

In LNSym, we already have a core language consisting of the `r` and `w` operators on `ARMState`.
We reflect precisely this subset of the `ARMState`, as well as arbitrary bitvector expressions for the value updates.
```

File: Tactics/ArmConstr.lean [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+  This is very barebones right now -- for instance, we'd need to support
+  `BitVec` operators here.
+-/
+inductive GPRVal where
+  | var (i : Nat)
+  | r_gpr (i : BitVec 5)
+  deriving DecidableEq, Repr, Inhabited
+
+abbrev StateContext := List ArmState
+abbrev GPRValContext := List (BitVec 64)
```
Comment (bollu) [inline Tactics/ArmConstr.lean:34]: I find the name a bit confusing, and I'm also not sure why these need to be looked up. Can't they directly be `(BitVec 64)`s that are reflected?

File: Tactics/ArmConstr.lean [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+      if i == gpr_idx then v else go gpr_idx rest
+
+/--
+Resolve any reads in each of `us` by looking them up in `es`.
+-/
+def Update.resolveReads (es us : List Update) : List Update :=
+  match us with
+  | [] => []
+  | u :: rest =>
+    (Update.resolveRead es u) :: resolveReads es rest
```
Comment (bollu) [inline Tactics/ArmConstr.lean:187]:
```suggestion
def Update.resolveReads (es us : List Update) : List Update := us.map (Update.resolveRead es)
  match us with
  | [] => []
  | u :: rest =>
    (Update.resolveRead es u) :: resolveReads es rest
```

**Comment by shigoel** [comment]
Closing PR; working on a different implementation.

**Pull request closed without merging**